sourcing-new/src/main/scala/ch/epfl/bluebrain/nexus/sourcingnew/projections/ProjectionStream.scala

```diff
@@ -0,0 +1,203 @@
+package ch.epfl.bluebrain.nexus.sourcingnew.projections
+
+import cats.effect.{ConcurrentEffect, Timer}
+import cats.implicits._
+import fs2.{Chunk, Stream}
+import io.circe.Encoder
+
+import scala.concurrent.duration.FiniteDuration
+import scala.util.control.NonFatal
+
+final case class IndexingConfig(batch: Int, batchTimeout: FiniteDuration)
+
+object ProjectionStream {
+
+  trait StreamOps[F[_], A] {
+
+    implicit def F: ConcurrentEffect[F]
+
+    protected def toResource[B, R](fetchResource: A => F[Option[R]], collect: R => Option[B])
+                        (implicit encoder : Encoder[A]): Message[A] => F[Message[B]] =
+      (message: Message[A]) =>
+        message match {
+          case s: SuccessMessage[A] =>
+            fetchResource(s.value)
+              .map {
+                _.flatMap(collect) match {
+                  case Some(value) =>
+                    message.map(_ => value)
+                  case _              => s.discarded
+                }
+              }
+              .recoverWith {
+                //TODO Handle errors properly
+                case NonFatal(err) => F.pure(s.failed(err))
+              }
+          case e: SkippedMessage      => F.pure(e)
+        }
+  }
+
+  implicit class SimpleStreamOps[F[_]: Timer, A: Encoder]
+    (val stream: Stream[F, Message[A]])(implicit override val F: ConcurrentEffect[F]) extends StreamOps[F, A] {
+
+    type O[_] = Message[_]
+
+    def resource[B, R](fetchResource: A => F[Option[R]], collect: R => Option[B]): Stream[F, Message[B]] =
+      stream.mapAsync(1)(toResource(fetchResource, collect))
```

Comment:
consider `evalMap` instead of `mapAsync(1)`. I don't exactly remember the issues with `mapAsync(1)` but I had several bad surprises when using it for the CLI

Reply:
Done